### DISCUSSION TITLE
Will HeteroData be supported in Batch?

### DISCUSSION THREAD

**Discussion by YAndrewL** [pyg-team/pytorch_geometric#3092] [Q&A] [answered]
Hi, I've been using torch_geometric.data.Batch to generate disconnected graph batches, but it seems to not work with HeteroData and raised error 'index out of range'. I'm wondering will the function will be supported with HeteroData in this version?

**Comment by rusty1s** [comment, marked as answer]
This should be supported, but it only got added in recently. Can you pull from master and see if that fixes your issues? Other-wise, please provide a minimal code snippet to reproduce so I can look into it.

**Reply by YAndrewL** [reply]
LOL it' an index problem, I've solved this, thanks a lot!

**Reply by rusty1s** [reply]
Awesome. Hope that everything works as expected now :)

**Reply by YAndrewL** [reply]
I have an extra question, how could I add a graph-level feature?

**Reply by rusty1s** [reply]
You can set it as an attribute of `data`, e.g.:
```python
data = HeteroData()
data['paper'].x = ...  # Node-level feature
data.y = ... # Graph-level feature

**Reply by YAndrewL** [reply]
What about a homogenous graph? Also directly give value to y?

**Reply by rusty1s** [reply]
What do you mean with homogeneous graphs? You can use the `torch_geometric.data.Data` class for storing homogeneous graphs. If you mean that you have connections between the same node type, you can directly hold them in `HeteroData` as well:
```python
data['paper', 'cites', 'paper'].edge_index = ...
```
Note that you can add any node-level, edge-level or graph-level attribute to `data` with any name.

**Reply by YAndrewL** [reply]
Here I mean if I have a homogeneous graph named data, and data.y records the label of each node and data.x records node feature. What I mean is I want to add a graph-level feature to this graph.
```python
data = Data()
data.x = ...  # this is node-level features
data.y = ...  # this is labels of each node,  how to add a graph-level feature?
````

**Reply by rusty1s** [reply]
The `data` object does not distinguish between node-level, edge-level or graph-level attributes. If you have a graph-level feature, you can just add it to `data` as well, e.g.:
```python
data.global_x = ...
```

**Reply by YAndrewL** [reply]
Got it! Very clear explanation, thanks a lot!

**Reply by YAndrewL** [reply]
Sorry 😭 Still have a question.
According to this example:
```python
from torch_geometric.nn import SAGEConv, to_hetero

class GNN(torch.nn.Module):
    def __init__(in_channels, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = SAGEConv(in_channels, hidden_channels)
        self.conv2 = SAGEConv(hidden_channels, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        x = self.conv2(x, edge_index)
        return x


model = GNN(in_channels=-1, hidden_channels=64, out_channels=dataset.num_classes)
model = to_hetero(model, data.metadata(), aggr='sum')
```

I simple substitute SAGEConv to GATConv, but it raised an error: `AttributeError: 'GATConv' object has no attribute 'lin_src'
`
Why did this happen?

**Reply by rusty1s** [reply]
Thanks for spotting and beta-testing our release :) This should be now fixed in master. I also tried to make it more clear in the documentation, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#automatically-converting-gnn-models). The issue is that we need to pass in a tuple as `in_channels` to each GNN operator to allow for message passing in bipartite graphs. You can create a heterogeneous graph attention network as follows:
```python
from torch_geometric.nn import GATConv, Linear, to_hetero

class GAT(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GATConv((-1, -1), hidden_channels, add_self_loops=False)
        self.lin1 = Linear(-1, hidden_channels)
        self.conv2 = GATConv((-1, -1), out_channels, add_self_loops=False)
        self.lin2 = Linear(-1, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index) + self.lin1(x)
        x = x.relu()
        x = self.conv2(x, edge_index) + self.lin2(x)
        return x


model = GAT(hidden_channels=64, out_channels=dataset.num_classes)
model = to_hetero(model, data.metadata(), aggr='sum')
```

**Reply by YAndrewL** [reply]
Thanks, I'm unfamiliar with the paper, is this model consistent with [this paper](https://arxiv.org/abs/1903.07293)?

**Reply by rusty1s** [reply]
The above example describes a variant of [R-GCN](https://arxiv.org/abs/1703.06103) in which the GNN operator is described via graph attention. The HAN operator is quite different, as it computes attention across metapaths and uses a semantic-level aggregation to combine representations along the metapath. We have plans to add HAN to PyG as well in a future release.
